README


This archive contains the distribution PlugAuth-Plugin-WebUI,
version 0.03:

  Embed a web user interface into your PlugAuth server

This software is copyright (c) 2012 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for {{$dist->name}}

{{$NEXT}}
  - Removed dependency on File::HomeDir

0.04      2015-09-10 11:22:03 -0400
  - Compatability with (and require) Mojo 6.x

0.03      2013-04-23 17:19:02 America/New_York
  - use MakeMaker instead of ModuleBuild
  - use [ACPS::RPM]
  - allow either simple_auth or plug_auth in config

0.02      2012-12-11 14:06:40 America/New_York
  - don't require credentials when SimpleAuth plugin not enabled

0.01      2012-12-07 15:47:09 America/New_York
  - initial version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::HomeDir" => 0,
    "File::HomeDir::Test" => 0,
    "Module::Build" => "0.3601",
    "Path::Class::Dir" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "YAML" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Embed a web user interface into your PlugAuth server",
  "dist_author" => [
    "Graham Ollis <gollis\@sesda3>"
  ],
  "dist_name" => "PlugAuth-Plugin-WebUI",
  "dist_version" => "0.02",
  "license" => "perl",
  "module_name" => "PlugAuth::Plugin::WebUI",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "PlugAuth" => "0.04",
    "PlugAuth::Role::Plugin" => 0,
    "PlugAuth::Role::Welcome" => 0,
    "PlugAuth::WebUI" => 0,
    "Role::Tiny::With" => 0,
    "perl" => "v5.10.0",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/PlugAuth/Plugin/WebUI.pm
package PlugAuth::Plugin::WebUI;

use strict;
use warnings;
use 5.010001;
use experimental qw( switch );
use PlugAuth::WebUI;
use Role::Tiny::With;

with 'PlugAuth::Role::Plugin';
with 'PlugAuth::Role::Welcome';

# ABSTRACT: Embed a web user interface into your PlugAuth server
# VERSION

=head1 SYNOPSIS

Your PlugAuth.conf

 ---
 url: http://localhost:3000
 plugins:
  - PlugAuth::Plugin::WebUI: {}

Start L<PlugAuth>

 % plugauth daemon

and then navigate to the PlugAuth WebUI:

 http://localhost:3000/ui

=head1 DESCRIPTION

This plugin embeds the L<PlugAuth WebUI|PlugAuth::WebUI> into your PlugAuth server,
which can be accessed via the url C</ui>.

=cut

sub init
{
  my($self) = @_;
  
  my $app = $self->app;
  
  my $share_dir = PlugAuth::WebUI->share_dir;
  my $data      = PlugAuth::WebUI->get_data;
  
  push @{ $app->renderer->paths}, $share_dir->subdir('tmpl')->stringify;
  
  $app->routes->route('/ui')->name('plugauth_webui')->get(sub {
    my($c) = @_;
    # FIXME: this comes out wrong when using an ssh tunnel.
    #        woraround was to set to '/' instead
    $data->{plugauth_webui_data}->{api_url} = $c->url_for('index')->to_abs->to_string;
    $data->{plugauth_webui_data}->{requires_authentic_credentials} = ($app->config->simple_auth(default => '') || $app->config->plug_auth(default => '')) ? 1 : 0;
    $c->stash($data);
    $c->render( template => 'plugauth_webui' );
  });
  
  $app->routes->route('/ui/:type/#file')->name('plugauth_webui_static')->get(sub {
    my($c) = @_;
    
    my $content_type;
    
    my $type = $c->param('type');
    
    given($type)
    {
      when('css') { $content_type = 'text/css'                 }
      when('js')  { $content_type = 'application/javascript'   }
      when('ico') { $content_type = 'image/vnd.microsoft.icon' }
      when('img') { $content_type = 'image/png';               }
      default { return $c->render_not_found }
    }
    
    my $file = $share_dir->file( $type, $c->param('file') );
    
    if(-r $file)
    {
      $c->res->headers->content_type($content_type);
      my $data = $file->slurp;
      $c->render(data => $data);
    }
    else
    {
      $c->render_not_found;
    }
  });
  
  if($self->plugin_config->test(default => ''))
  {
    $app->routes->route('/t/:test')->name('plugauth_webui_test')->get(sub {
      my($c) = @_;
      my $test = $c->param('test');
      my $file = $share_dir->file( 't', "$test.js" );
      return $c->render_not_found unless -r $file;
      
      $data->{plugauth_webui_data}->{api_url} = $c->url_for('index')->to_abs;
      $data->{plugauth_webui_data}->{requires_authentic_credentials} = ($app->config->simple_auth(default => '') || $app->config->plug_auth(default => '')) ? 1 : 0;
      $c->stash($data);
      
      $c->stash->{test} = $test;
      if($c->stash->{format} eq 'js')
      {
        $c->res->headers->content_type('application/x-javascript');
        my $data = $file->slurp;
        $c->render(data => $data);
      }
      else
      {
        $c->render( template => 'plugauth_webui_test' );
      }
    });
    
    $app->routes->route('/t')->name('plugauth_webui_test_list')->get(sub {
      my($c) = @_;
      $c->stash->{list} = [ grep { s/\.js$// } map { $_->basename } $share_dir->subdir('t')->children(no_hidden => 1) ];
      $c->render( template => 'plugauth_webui_test_list' );
    });
  }
}

sub welcome
{
  my($self, $c) = @_;
  $c->redirect_to($c->url_for('plugauth_webui'));
}

1;


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Embed a web user interface into your PlugAuth server",
  "AUTHOR" => "Graham Ollis <gollis\@sesda3>",
  "BUILD_REQUIRES" => {
    "File::HomeDir" => 0,
    "File::HomeDir::Test" => 0,
    "Path::Class::Dir" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "YAML" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "PlugAuth-Plugin-WebUI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "PlugAuth::Plugin::WebUI",
  "PREREQ_PM" => {
    "PlugAuth" => "0.04",
    "PlugAuth::Role::Plugin" => 0,
    "PlugAuth::Role::Welcome" => 0,
    "PlugAuth::WebUI" => 0,
    "Role::Tiny::With" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



